File: examples/crypto-tracker/sdks/c/src/services.c
#include "services.h"
#include <stdlib.h>
#include <string.h>

app_response_t* app_coin_gecko_get_coin_detail(app_client_t* client, const app_request_opts_t* opts) {
    app_request_opts_t local_opts = {0};
    if (opts) local_opts = *opts;
    local_opts.method = "GET";
    local_opts.endpoint = "coins/{id}";
    return app_client_request(client, &local_opts);
}

app_response_t* app_coin_gecko_get_global_data(app_client_t* client, const app_request_opts_t* opts) {
    app_request_opts_t local_opts = {0};
    if (opts) local_opts = *opts;
    local_opts.method = "GET";
    local_opts.endpoint = "global";
    return app_client_request(client, &local_opts);
}

app_response_t* app_coin_gecko_get_market_chart(app_client_t* client, const app_request_opts_t* opts) {
    app_request_opts_t local_opts = {0};
    if (opts) local_opts = *opts;
    local_opts.method = "GET";
    local_opts.endpoint = "coins/{id}/market_chart";
    return app_client_request(client, &local_opts);
}

app_response_t* app_coin_gecko_get_top_coins(app_client_t* client, const app_request_opts_t* opts) {
    app_request_opts_t local_opts = {0};
    if (opts) local_opts = *opts;
    local_opts.method = "GET";
    local_opts.endpoint = "coins/markets";
    return app_client_request(client, &local_opts);
}

app_response_t* app_coin_gecko_get_trending_coins(app_client_t* client, const app_request_opts_t* opts) {
    app_request_opts_t local_opts = {0};
    if (opts) local_opts = *opts;
    local_opts.method = "GET";
    local_opts.endpoint = "search/trending";
    return app_client_request(client, &local_opts);
}

app_response_t* app_coin_gecko_search_coins(app_client_t* client, const app_request_opts_t* opts) {
    app_request_opts_t local_opts = {0};
    if (opts) local_opts = *opts;
    local_opts.method = "GET";
    local_opts.endpoint = "search";
    return app_client_request(client, &local_opts);
}

File: examples/crypto-tracker/sdks/rust/src/services.rs
use crate::client::Client;
use reqwest::header::HeaderMap;
use std::collections::HashMap;

/// Service for CoinGeckoService containing 6 endpoints
pub struct CoinGeckoService;

impl CoinGeckoService {
    /// GET coins/{id}
    pub async fn get_coin_detail(
        client: &Client,
        id: &str,
        query_params: Option<&HashMap<&str, &str>>,
    ) -> Result<reqwest::Response, reqwest::Error> {
        let mut path_params = HashMap::new();
        path_params.insert("id", id);
        let pp_arg = Some(&path_params);
        client.send_request(
            reqwest::Method::GET,
            "coins/{id}",
            pp_arg,
            query_params,
            None,
            None,
        ).await
    }

    /// GET global
    pub async fn get_global_data(
        client: &Client,
    ) -> Result<reqwest::Response, reqwest::Error> {
        let pp_arg = None;
        client.send_request(
            reqwest::Method::GET,
            "global",
            pp_arg,
            None,
            None,
            None,
        ).await
    }

    /// GET coins/{id}/market_chart
    pub async fn get_market_chart(
        client: &Client,
        id: &str,
        query_params: Option<&HashMap<&str, &str>>,
    ) -> Result<reqwest::Response, reqwest::Error> {
        let mut path_params = HashMap::new();
        path_params.insert("id", id);
        let pp_arg = Some(&path_params);
        client.send_request(
            reqwest::Method::GET,
            "coins/{id}/market_chart",
            pp_arg,
            query_params,
            None,
            None,
        ).await
    }

    /// GET coins/markets
    pub async fn get_top_coins(
        client: &Client,
        query_params: Option<&HashMap<&str, &str>>,
    ) -> Result<reqwest::Response, reqwest::Error> {
        let pp_arg = None;
        client.send_request(
            reqwest::Method::GET,
            "coins/markets",
            pp_arg,
            query_params,
            None,
            None,
        ).await
    }

    /// GET search/trending
    pub async fn get_trending_coins(
        client: &Client,
    ) -> Result<reqwest::Response, reqwest::Error> {
        let pp_arg = None;
        client.send_request(
            reqwest::Method::GET,
            "search/trending",
            pp_arg,
            None,
            None,
            None,
        ).await
    }

    /// GET search
    pub async fn search_coins(
        client: &Client,
        query_params: Option<&HashMap<&str, &str>>,
    ) -> Result<reqwest::Response, reqwest::Error> {
        let pp_arg = None;
        client.send_request(
            reqwest::Method::GET,
            "search",
            pp_arg,
            query_params,
            None,
            None,
        ).await
    }

}
